userspace/libscap/clock_helpers.h

```diff
@@ -0,0 +1,85 @@
+/*
+Copyright (C) 2022 The Falco Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+*/
+#ifndef __CLOCK_HELPERS_H
+#define __CLOCK_HELPERS_H
+
+#define SCAP_GET_CUR_TS_MS_CONTEXT_INIT ((uint64_t)0)
+#define SCAP_GET_CUR_TS_MS_CONTEXT_ERROR_FLAG ((uint64_t)0x8000000000000000)
+#define SCAP_GET_CUR_TS_MS_CONTEXT_PREV_VALUE_MASK ((uint64_t)0x7fffffffffffffff)
+
+#define S_TO_MS(_sec) (((uint64_t)_sec) * (uint64_t)1000)
+#define NS_TO_MS(_ns) (((uint64_t)_ns) / ((uint64_t)(1000 * 1000)))
+
+#ifndef __always_inline
+#define __always_inline inline
+#endif
+
+/**
+ * Return monotonically increasing time in ms.
+ * Caller initializes context to SCAP_GET_CUR_TS_MS_CONTEXT_INIT,
+ * Function uses and updates context, to recognize and handle the
+ * following cases:
+ * - failed clock_gettime() system call
+ * - non-monotonic behavior of CLOCK_MONOTONIC
+ * - time values that cannot be represented in uint64_t number of msec
+ */
+static __always_inline uint64_t scap_get_monotonic_ts_ms(uint64_t* context)
+{
+	// Record previously reported time; will be 0 for first call.
+	uint64_t prev_time = ((*context) & SCAP_GET_CUR_TS_MS_CONTEXT_PREV_VALUE_MASK);
+
+	// If context indicates error already detected, just return the
+	// last reported time
+	if ((*context) & SCAP_GET_CUR_TS_MS_CONTEXT_ERROR_FLAG)
```

Comment:
if we do `((*context) & SCAP_GET_CUR_TS_MS_CONTEXT_PREV_VALUE_MASK);` at the previous line this condition would be always false, or am I missing something?

Reply:
🤔 we're not changing `*context` here (or in line 43). This might be clearer with bitfields:
```
uint64_t error:1;
uint64_t prev_time:63;
```

but I'm irrationally scared of bitfields ;)

Reply:
sorry I'm stupid :facepalm: it's perfectly fine!